Refactor TLS1-PRF to create the MAC contexts early

levitte opened the pull request:
The TLS1-PRF implementation fetched the digest(s) for the underlying
MAC, just to get their names and pass those down to the MAC, which in
turn would fetch those same digests again.

This change circumvents this by fetching the MAC (or MACs in the
MD5-SHA1 special case) and create the MAC contexts for them directly
when this PRF receives the relevant parameters, thus only having to
pass EVP_MAC_CTX pointers around.

levitte commented:
This is really a remnant from pre-3.0, where EVP_MDs and EVP_CIPHERs are passed around rather than their names.  We obviously sometimes still want to pass around EVP_MDs and EVP_CIPHERs, sometimes unnecessarily.  Our provider code could do well with a bit of extra scrutiny to see if we can catch this old habit.

paulidale commented on providers/common/kdfs/tls1_prf.c:
```diff
@@ -276,23 +310,31 @@ static int tls1_prf_P_hash(const EVP_MD *md,
     size_t Ai_len;
     int ret = 0;
-    OSSL_PARAM params[4];
+    OSSL_PARAM params[10], *p = params;
```
[6] is required here, not [10]

levitte replied:
Yeah, I set 10 to be safe before i started modifying and forgot to adjust.  Will fix.

paulidale commented on providers/common/kdfs/tls1_prf.c:
```diff
@@ -78,10 +79,18 @@ static int tls1_prf_alg(const EVP_MD *md, const EVP_MD *sha1,
+    char *digest;
     /* Second digest for the MD5/SHA-1 combined PRF */
-    PROV_DIGEST sha1;
+    const char *sha1;
```
Why?  This one is constant?

Resolving early is a performance gain.

levitte replied:
Good question, it can be hard coded in `tls1_prf_alg`...

paulidale reviewed (commented):
I'm more leaning towards a `PROV_MAC` structure and associated helpers.  Early resolution of algorithms is good.

Other KDFs/PRFs would benefit from `PROC_MAC` (SSKDF uses a mac and KBKDF in #9924 does too).

I've likely missed some subtleties like I missed the `prov_ctx`.

levitte commented:
> Early resolution of algorithms is good.

Well, in this case, they are getting "resolved", just to get "resolved" again by the MAC!  That's utterly ridiculous, and only a waste of cycles.

paulidale commented:
> Well, in this case, they are getting "resolved", just to get "resolved" again by the MAC! That's utterly ridiculous, and only a waste of cycles.

My poor choice of wording.  The MAC and any subservient algorithms should be resolved at _set_params time.

levitte commented:
> My poor choice of wording. The MAC and any subservient algorithms should be resolved at `_set_params` time.

Aaahhhhhh!  Now that's a good idea!

I need a bit of time to get that right...

levitte commented:
However, there's no need for a `PROV_MAC` structure since we don't need to deal with anything legacy there...

paulidale commented:
> However, there's no need for a PROV_MAC structure since we don't need to deal with anything legacy there...

Do the underlying algorithms used alter this?
I.e. should the PROV_MAC structure include PROC_DIGEST and PROV_CIPHER fields for the MACS?
Perhaps a union of the two.

levitte commented:
Hold ye' 'orses, @paulidale :wink:

levitte committed:
Refactor TLS1-PRF to create the MAC contexts early

The TLS1-PRF implementation fetched the digest(s) for the underlying
MAC, just to get their names and pass those down to the MAC, which in
turn would fetch those same digests again.

This change circumvents this by fetching the MAC (or MACs in the
MD5-SHA1 special case) and create the MAC contexts for them directly
when this PRF receives the relevant parameters, thus only having to
pass EVP_MAC_CTX pointers around.

levitte commented:
Okie, I threw away the previous change and started over.  Please have a new look.

t8m commented on providers/common/kdfs/tls1_prf.c:
```diff
@@ -137,38 +138,73 @@ static int kdf_tls1_prf_derive(void *vctx, unsigned char *key,
                         key, keylen);
 }
 
+static EVP_MAC_CTX *kdf_tls1_prf_mkmac(OPENSSL_CTX *libctx,
```
Nit: The mkmac name seems slightly confusing to me as it creates MAC_CTX and not the actual mac. So I'd prefer something like kdf_tls1_prf_macctx or kdf_tls1_prf_mkmacctx. 
But this is just for your consideration the approval holds even without the change.

levitte replied:
I agree with you, so I'll rename

levitte committed:
fixup! Refactor TLS1-PRF to create the MAC contexts early

levitte commented:
Merged.

9a92bf1bffad15ede5ac97d1f1705c3e2c249a98 Refactor TLS1-PRF to create the MAC contexts early

levitte commented:
Thanks all!
